gh-124889: Remove redundant artificial rules in PEG parser

efimov-mikhail opened the pull request:
Cache in C PEG-generator reworked:
we save artificial rules in cache by Node string representation as a key instead of Node object itself. As a result total count of artificial rules in parsers.c is lowered from 283 to 170. More natural number ordering is used for the names of artificial rules.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-124889 -->
* Issue: gh-124889
<!-- /gh-issue-number -->

efimov-mikhail committed:
gh-124889: Remove redundant artificial rules in PEG parser

Cache in C PEG-generator reworked:
we save artificial rules in cache by Node string representation as a key instead of Node object itself.
As a result total count of artificial rules in parsers.c is lowered from 283 to 170.
More natural number ordering is used for the names of artificial rules.

pablogsal commented on Tools/peg_generator/pegen/c_generator.py:
```diff
@@ -207,19 +207,23 @@ def visit_StringLeaf(self, node: StringLeaf) -> FunctionCall:
-            self.cache[node] = self.generate_call(node.alts[0].items[0])
+            return self.generate_call(node.alts[0].items[0])
+
+        node_str = f"{node}"
```
It would be great if we can abstract this away in a function. Maybe

```
def generate_artificial_rule(node, prefix, new_rule_fn):
    node_str = f"{node}"
    key = f"{prefix}_{node_str}"
    if key in self.cache:
             name = self.cache[key]
     else:
             name = new_rule_fn()
             self.cache[key] = name
       return name
```

pablogsal reviewed (approved):
LGTM! Great find

Left a small nit. Tell me what you think

efimov-mikhail committed:
gh-124889: Remove redundant artificial rules, some refactoring (#124893)

Auxiliary method CCallMakerVisitor._generate_artificial_rule_call is added.
Its purpose is abstracting work with artificial rules cache.

lysnikolaou commented on Tools/peg_generator/pegen/c_generator.py:
```diff
@@ -302,44 +287,62 @@ def visit_Opt(self, node: Opt) -> FunctionCall:
+    ) -> FunctionCall:
+        node_str = f"{node}"
+        key = f"{prefix}_{node_str}"
+        if key in self.cache:
+            name = self.cache[key]
+        else:
+            name = rule_generation_func()
+            self.cache[key] = name
```
We can use `setdefault` for this.

```suggestion
        name = self.cache.setdefault(key, rule_generation_func())
```

efimov-mikhail replied:
Unfortunately, this change is not correct: rule_generation_func will be called even in the case when key is already present in dict.
But rule_generation_func is not an idempotent function, and redundant rules will be generated.

On the same reason we can't write something like this:
name = cache.get(key, rule_generation_func())
cache[key] = name

pablogsal replied:
The difference here is that the rule generator function is always called even if the element is in the cache

lysnikolaou replied:
Hmm, you're both right, of course. I missed that.

lysnikolaou commented on Tools/peg_generator/pegen/c_generator.py:
```diff
@@ -302,44 +287,62 @@ def visit_Opt(self, node: Opt) -> FunctionCall:
+        return self._generate_artificial_rule_call(
+            node,
+            "repeat0",
+            lambda: self.gen.artificial_rule_from_repeat(node.node, False),
```
Can we use the kwarg here so that it's clear what `True` or `False` is for? Same goes for the call below.

```suggestion
            lambda: self.gen.artificial_rule_from_repeat(node.node, is_repeat1=False),
```

efimov-mikhail replied:
Totally agree with this change.

Maybe we should provide similar change in python_generator.py file?
Cache rework in general will be useful for python_generator too, I suppose

pablogsal replied:
Let's do that in a separate PR to keep the changes here minimum

efimov-mikhail replied:
Of course, separate PR is a great choice

lysnikolaou reviewed (commented):
Thanks for this @efimov-mikhail! It's a very nice improvement. I've left a couple of more minor comments about the code.

efimov-mikhail committed:
gh-124889: Remove redundant artificial rules, some refactoring 2 (#124893)

Explicit using of "is_repeat1" kwarg is added to visit_Repeat0 and visit_Repeat1 methods.
Its slightly improve code readabitily.

efimov-mikhail committed:
Merge branch 'main' into peg_parser_remove_redundant_functions

lysnikolaou reviewed (approved):
LGTM!
